tensorboard/backend/event_processing/plugin_event_multiplexer_test.py

```diff
@@ -295,30 +295,97 @@ def testAddRunMaintainsLoading(self):
+    self.assertTrue(x.GetAccumulator('run1').reload_called)
+    self.assertTrue(x.GetAccumulator('run2').reload_called)
+    self.assertTrue(x.GetAccumulator('run3').reload_called)
+
 
 class EventMultiplexerWithRealAccumulatorTest(tf.test.TestCase):
 
   def testDeletingDirectoryRemovesRun(self):
     x = event_multiplexer.EventMultiplexer()
     tmpdir = self.get_temp_dir()
-    join = os.path.join
-    run1_dir = join(tmpdir, 'run1')
-    run2_dir = join(tmpdir, 'run2')
-    run3_dir = join(tmpdir, 'run3')
+    self.add3RunsToMultiplexer(tmpdir, x)
+    x.Reload()
 
-    for dirname in [run1_dir, run2_dir, run3_dir]:
-      _AddEvents(dirname)
+    # Delete the directory, then reload.
+    shutil.rmtree(os.path.join(tmpdir, 'run2'))
+    x.Reload()
+    self.assertNotIn('run2', x.Runs().keys())
+
+  def testReloadWithMoreRunsThanThreads(self):
+    patcher = tf.test.mock.patch('threading.Thread.start', autospec=True)
+    start_mock = patcher.start()
+    self.addCleanup(patcher.stop)
+    patcher = tf.test.mock.patch(
+        'six.moves.queue.Queue.join', autospec=True)
+    join_mock = patcher.start()
+    self.addCleanup(patcher.stop)
 
-    x.AddRun(run1_dir, 'run1')
-    x.AddRun(run2_dir, 'run2')
-    x.AddRun(run3_dir, 'run3')
+    x = event_multiplexer.EventMultiplexer(max_reload_threads=2)
+    tmpdir = self.get_temp_dir()
+    self.add3RunsToMultiplexer(tmpdir, x)
+    x.Reload()
+
+    # 2 threads should have been started despite how there are 3 runs.
+    self.assertEqual(2, start_mock.call_count)
+    self.assertEqual(1, join_mock.call_count)
+
+  def testReloadWithMoreThreadsThanRuns(self):
+    patcher = tf.test.mock.patch('threading.Thread.start', autospec=True)
+    start_mock = patcher.start()
+    self.addCleanup(patcher.stop)
+    patcher = tf.test.mock.patch(
+        'six.moves.queue.Queue.join', autospec=True)
+    join_mock = patcher.start()
+    self.addCleanup(patcher.stop)
 
+    x = event_multiplexer.EventMultiplexer(max_reload_threads=42)
+    tmpdir = self.get_temp_dir()
+    self.add3RunsToMultiplexer(tmpdir, x)
     x.Reload()
 
-    # Delete the directory, then reload.
-    shutil.rmtree(run2_dir)
+    # 3 threads should have been started despite how the multiplexer
+    # could have started up to 42 threads.
+    self.assertEqual(3, start_mock.call_count)
+    self.assertEqual(1, join_mock.call_count)
+
+  def testReloadWith1Thread(self):
+    patcher = tf.test.mock.patch('threading.Thread.start', autospec=True)
+    start_mock = patcher.start()
+    self.addCleanup(patcher.stop)
+    patcher = tf.test.mock.patch(
+        'six.moves.queue.Queue.join', autospec=True)
+    join_mock = patcher.start()
+    self.addCleanup(patcher.stop)
+
+    x = event_multiplexer.EventMultiplexer(max_reload_threads=1)
+    tmpdir = self.get_temp_dir()
+    self.add3RunsToMultiplexer(tmpdir, x)
     x.Reload()
-    self.assertNotIn('run2', x.Runs().keys())
+
+    # The multiplexer should have started no new threads.
```

Comment:
Would it make sense to also assert on the actual multiplexer contents here to make sure the Reload() actually worked?  Since the single-thread reload should run synchronously, it should work normally.

Reply:
We now use mock accumulators. There are tests in the other suite that check for correctness based on whether the right accumulators reloaded (that mocked reload method is called).